pkg/xds/context/mesh_snapshot.go

```diff
@@ -20,46 +20,66 @@ import (
 
 var meshCacheLog = core.Log.WithName("mesh-cache")
 
-// meshSnapshot represents all resources that belong to Mesh and allows to calculate hash.
+// MeshSnapshot represents all resources that belong to Mesh and allows to calculate hash.
 // Calculating and comparing hashes is much faster than call 'equal' for xDS resources. So
 // meshSnapshot reduces costs on reconciling Envoy config when resources in the store are
 // not changing
-type meshSnapshot struct {
+type MeshSnapshot struct {
 	mesh      *core_mesh.MeshResource
 	resources map[core_model.ResourceType]core_model.ResourceList
 	ipFunc    lookup.LookupIPFunc
 }
 
-func GetMeshSnapshot(ctx context.Context, meshName string, rm manager.ReadOnlyResourceManager, types []core_model.ResourceType, ipFunc lookup.LookupIPFunc) (*meshSnapshot, error) {
-	snapshot := &meshSnapshot{
+type MeshSnapshotBuilder interface {
+	Build(ctx context.Context, meshName string) (*MeshSnapshot, error)
+}
+
+func NewMeshSnapshotBuilder(rm manager.ReadOnlyResourceManager, types []core_model.ResourceType, ipFunc lookup.LookupIPFunc) MeshSnapshotBuilder {
+	return &meshSnapshotBuilder{
+		rm:     rm,
+		types:  types,
+		ipFunc: ipFunc,
+	}
+}
+
+type meshSnapshotBuilder struct {
+	rm     manager.ReadOnlyResourceManager
+	types  []core_model.ResourceType
+	ipFunc lookup.LookupIPFunc
+}
+
+var _ MeshSnapshotBuilder = &meshSnapshotBuilder{}
+
+func (m *meshSnapshotBuilder) Build(ctx context.Context, meshName string) (*MeshSnapshot, error) {
+	snapshot := &MeshSnapshot{
 		resources: map[core_model.ResourceType]core_model.ResourceList{},
-		ipFunc:    ipFunc,
+		ipFunc:    m.ipFunc,
 	}
 
 	mesh := core_mesh.NewMeshResource()
-	if err := rm.Get(ctx, mesh, core_store.GetByKey(meshName, core_model.NoMesh)); err != nil {
+	if err := m.rm.Get(ctx, mesh, core_store.GetByKey(meshName, core_model.NoMesh)); err != nil {
 		return nil, err
 	}
 	snapshot.mesh = mesh
 
-	for _, typ := range types {
+	for _, typ := range m.types {
 		switch typ {
 		case core_mesh.DataplaneType:
```

Comment:
I'm wondering why do we have special cases for `DataplaneType` and `ZoneIngressType`. Maybe it was required when Ingress and Dataplane had the same type, but now it feels like it could be handled by `default` section of the `switch`

Reply:
ZoneIngress makes sense because we do not pass `ListByMesh`, Dataplane can be removed indeed